### PR TITLE
Oneof fix

### PR DESCRIPTION
**Pull request by aluu317**
**What this PR does / why we need it**:

The previous code relies on the existence of `_sequence` in the field name for union of list types, or splitting on underscores `_` to get the oneof name (how silly!). 

This PR:
- removes the check for `_sequence` because users could name their oneof field names custom names without the word `sequence` and without `_`.
- lots of tests to cover the cases of default union of list names when users don't provide a field name (ex: "foo_int_sequence") as well as when they provide custom names (ex: "myfooseq")

**Special notes for your reviewer**:

**If applicable**:
- [ ] this PR contains documentation
- [ ] this PR contains unit tests
- [ ] this PR has been tested for backwards compatibility

### CODE REVIEW

File: caikit/core/data_model/base.py [hunk 1]
```diff
@@ -666,12 +669,8 @@ def from_proto(cls, proto):
                 )
 
             if field in cls._fields_primitive or field in cls._fields_enum:
-                if field in cls._fields_to_oneof:
-                    if proto.HasField(field):
-                        # "foo_bar_int" has original field name "foo_bar"
-                        original_field_name = "_".join(field.split("_")[:-1])
-                        kwargs[original_field_name] = proto_attr
-                else:
+                # special case for oneofs
```
Comment (aluu317) [inline caikit/core/data_model/base.py:672]: This reverts the changes introduced by https://github.com/caikit/caikit/pull/350/files. We don't need this change

File: tests/core/data_model/test_dataobject.py [hunk 2]
```diff
@@ -575,6 +634,101 @@ class Foo(DataObjectBase):
     assert Foo.from_json(json_repr_foo) == foo1
 
 
+def test_dataobject_list_oneof_round_trips_with_specified_names():
+    @dataobject
+    class Foo(DataObjectBase):
+        foo: Union[
+            Annotated[List[int], FieldNumber(10), OneofField("myfoointseq")],
+            Annotated[List[str], FieldNumber(20), OneofField("myfoostrseq")],
```
Comment (joerunde) [inline tests/core/data_model/test_dataobject.py:642]: quick question: What's different between this test and the one above that has
```
    class Foo(DataObjectBase):
        foo: Union[
            Annotated[int, FieldNumber(10), OneofField("myfooint")],
            Annotated[float, FieldNumber(20), OneofField("myfoofloat")],
        ]
```

I can't quickly tell if the only difference here is the names or if there's something else different going on

**Review by joerunde — APPROVED** [review]
Tests look nice to me, just had one question

File: tests/core/data_model/test_dataobject.py (hunk 2 above)
Comment (aluu317) [reply tests/core/data_model/test_dataobject.py:642]: This one is union of lists, the above is union of int and float

**Pull request merged**